src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -465,5 +481,25 @@ public virtual MethodDesc GetFinalizer()
         {
             return null;
         }
+
+        /// <summary>
+        /// Gets a value indicating whether this type has generic variance (the definition of the type
+        /// has a generic parameter that is co- or contravariant).
+        /// </summary>
+        public bool HasVariance
+        {
+            get
+            {
+                return (GetTypeFlags(TypeFlags.HasGenericVariance | TypeFlags.HasGenericVarianceComputed) & TypeFlags.HasGenericVariance) != 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the kind of this type.
+        /// </summary>
+        public abstract TypeKind Variety
```

Comment:
I think it is good enough that the overall structure is similar. I do not think we need to make out of our way to make it same. I would like for this one to be best casting implementation that we got - easy to understand, fast, complete. It is what my comments were about.

After looking into details, I actually like the overall structure of this algorithm better than the one in `Runtime.Base\src\System\Runtime\TypeCast.cs`. I think the unification may rather go in the other direction. And  we should be able to replace the copies of the casting algorithm in reflection with this one eventually. There are actually two after taking a closer look - one in `System.Private.Reflection.Execution\src\Internal\Reflection\Execution\TypeLoader\TypeCast.cs` and other one in `System.Private.Reflection.Core\System\Reflection\Runtime\General\Assigninibity.cs`.